### DISCUSSION TITLE
How to add DNS hosted on VPN server to clients configuration

### DISCUSSION THREAD

**Discussion by fancywriter** [strongswan/strongswan#1413] [Q&A]
Hello! I am trying to understand how virtual IPs work. https://docs.strongswan.org/docs/6.0/features/vip.html

I have this simple configuration and it works exactly as I expect:
```
connections {
    conn_eap {
        version = 2
        pools = pool
        local {
            certs = cert.pem
            id = <FQDN>
        }
        remote {
            id = %any
            auth = eap-mschapv2
        }
        children {
            conn {
                local_ts = 0.0.0.0/0
            }
        }
        send_cert = always # because not all clients request this cert explicitly
        unique = never # this is to add different IPs for clients sharing same password, for simplicity
        proposals = default,aes256-sha256-modp1024 # this is for Windows clients, by default modp1024 is turned off
    }
}

secrets {
    eap-user {
        secret = password
        id = user
    }
}

pools {
    pool {
        addrs = 10.10.10.0/24
    }
}
```

If I add constant DNS nameservers to the pools like this for example:
```
pools {
    pool {
        addrs = 10.10.10.0/24
        dns = 94.140.14.14, 94.140.15.15 # adguard DNS
    }
}
```

Works fine. However, I would like to have more control over my DNS, I would like to use dnsmasq (to enable some hosts banned by adguard and disable some others not banned by them let's say). And for simplicity I am going to host it on the same server as VPN. I have installed dnsmasq, enabled port 53 in firewall and tested that DNS lookup  (`nslookup`) work through my server.

However, I don't have static IP. Am I required to have static IP to make it work? Or is it possible with dynamic IP? Is it possible to assign virtual IP to the VPN server itself? By the way, it doesn't have to serve DNS itself, it can forward traffic on port 53 to anywhere I configure firewall...

I have the pool 10.10.10.0/24
```
$ ipcalc 10.10.10.0/24 | grep HostM
HostMin:   10.10.10.1           00001010.00001010.00001010. 00000001
HostMax:   10.10.10.254         00001010.00001010.00001010. 11111110
```

And it assigned IPs 10.10.10.1, 10.10.10,2, ... etc to each new client. Can I reserve 10.10.10.1 (or 10.10.10.127) to VPN server hosting DNS (or anything else) and allow traffic on port 53 on it and add configure it like this?

```
pools {
    pool {
        addrs = 10.10.10.0/25 # 10.10.10.1 - 10.10.10.126
        dns = 10.10.10.127
    }
}
```

Or is using something like [own DHCP](https://docs.strongswan.org/docs/6.0/plugins/dhcp.html) server is the only option?

**Comment by fractalb** [comment]
> pools {
>     pool {
>         addrs = 10.10.10.0/25 # 10.10.10.1 - 10.10.10.126
>         dns = 10.10.10.127
>     }
> }


Exclude the 10.10.10.127 (DNS) address from the pool addresses and it should work, I think.
addrs  = 10.10.10.1 - 10.10.10.126
dns  = 10.10.10.127
But there won't be any automatic allocation of the DNS ip to your server. You need to assign the DNS IP to one of the interfaces on your server manually, possibly by creating a new virtual interface.

**Comment by tobiasbrunner** [comment]
> Can I reserve 10.10.10.1 (or 10.10.10.127) to VPN server hosting DNS (or anything else) and allow traffic on port 53 on it and add configure it like this?

Yes, if you configure `addrs = 10.10.10.2/25`, the first address assigned to clients will be `10.10.10.2` and you can assign `10.10.10.1` to `lo` of the server and use that as DNS server address.
